presto-main/src/main/java/com/facebook/presto/execution/PageBuffer.java

```diff
@@ -0,0 +1,202 @@
+        }
+        else {
+            QueuedPage queuedPage = new QueuedPage(page);
+            queuedPages.add(queuedPage);
+            return queuedPage.getFuture();
+        }
+    }
+
+    private synchronized void addToMasterBuffer(Page page)
+    {
+        long bytesAdded = 0;
+        List<Page> pages = splitPage(page, DEFAULT_MAX_PAGE_SIZE_IN_BYTES);
+        masterBuffer.addAll(pages);
+        pagesAdded.addAndGet(pages.size());
+        for (Page p : pages) {
+            bytesAdded += p.getSizeInBytes();
+        }
+        bufferedBytes.addAndGet(bytesAdded);
+    }
+
+    public synchronized List<Page> getPages(DataSize maxSize, long sequenceId)
+    {
+        long maxBytes = maxSize.toBytes();
+        List<Page> pages = new ArrayList<>();
+        long bytes = 0;
+
+        int listOffset = Ints.checkedCast(sequenceId - masterSequenceId.get());
+        while (listOffset < masterBuffer.size()) {
+            Page page = masterBuffer.get(listOffset++);
+            bytes += page.getSizeInBytes();
+            // break (and don't add) if this page would exceed the limit
+            if (!pages.isEmpty() && bytes > maxBytes) {
+                break;
+            }
+            pages.add(page);
+        }
+        return ImmutableList.copyOf(pages);
+    }
+
+    public synchronized void advanceSequenceId(long newSequenceId)
+    {
+        long oldMasterSequenceId = masterSequenceId.get();
+        checkArgument(newSequenceId >= oldMasterSequenceId, "Master sequence id moved backwards: oldMasterSequenceId=%s, newMasterSequenceId=%s",
+                oldMasterSequenceId,
+                newSequenceId);
+
+        if (newSequenceId == masterSequenceId.get()) {
+            return;
+        }
+        masterSequenceId.set(newSequenceId);
+
+        // drop consumed pages
+        int pagesToRemove = Ints.checkedCast(newSequenceId - oldMasterSequenceId);
+        checkState(masterBuffer.size() >= pagesToRemove,
+                "MasterBuffer does not have any pages to remove: pagesToRemove %s oldMasterSequenceId: %s newSequenceId: %s",
+                pagesToRemove,
+                oldMasterSequenceId,
+                newSequenceId);
+        for (int i = 0; i < pagesToRemove; i++) {
+            Page page = masterBuffer.removeFirst();
+            this.bufferedBytes.addAndGet(-page.getSizeInBytes());
+        }
+        // refill buffer from queued pages
+        while (!queuedPages.isEmpty() && bufferedBytes.get() < maxBufferedBytes) {
+            QueuedPage queuedPage = queuedPages.remove();
+            addToMasterBuffer(queuedPage.getPage());
+            queuedPage.getFuture().set(null);
+        }
+    }
+
+    public synchronized void clear(boolean freeReaders)
+    {
+        // clear the buffer
+        masterBuffer.clear();
+        bufferedBytes.set(0);
+        if (freeReaders) {
+            // free queued page waiters
+            clearQueue();
+        }
+    }
+
+    public synchronized void clearQueue()
```

Comment:
I'd rename this something like `clearWriteQueue` or `dropPendingPages`... it isn't clear in the caller code what is going on especially with another method named `clear`